File: services/drakon-agent/routes/feedback.py
"""POST /feedback — persist diagram feedback into knowledge base for BM25 reindex."""
import hashlib
import json
import os
from datetime import datetime, timezone
from pathlib import Path

from fastapi import APIRouter
from pydantic import BaseModel

router = APIRouter()

KNOWLEDGE_DIR = Path(__file__).parent.parent / "knowledge"
FEEDBACK_DIR = KNOWLEDGE_DIR / "feedback"


class FeedbackRequest(BaseModel):
    diagram_name: str
    feedback: str
    corrected_ir: dict | None = None


def _feedback_to_markdown(req: FeedbackRequest) -> str:
    ts = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
    lines = [
        f"## Feedback: {req.diagram_name}",
        f"",
        f"**Date:** {ts}",
        f"**Diagram:** `{req.diagram_name}`",
        f"",
        f"### What was wrong",
        f"{req.feedback}",
    ]
    if req.corrected_ir:
        lines += [
            f"",
            f"### Corrected IR",
            f"```json",
            json.dumps(req.corrected_ir, ensure_ascii=False, indent=2),
            f"```",
            f"",
            f"**Key corrections:**",
        ]
        items = req.corrected_ir.get("items", {})
        for node_id, node in items.items():
            if node.get("type") == "question":
                lines.append(f"- `{node_id}`: question→one={node.get('one')}, two={node.get('two')}")
            elif node.get("type") == "action":
                lines.append(f"- `{node_id}`: action content=`{node.get('content','')[:60]}`")
    return "\n".join(lines)


@router.post("/feedback")
def feedback(req: FeedbackRequest):
    FEEDBACK_DIR.mkdir(parents=True, exist_ok=True)

    slug = hashlib.md5(f"{req.diagram_name}{req.feedback}".encode()).hexdigest()[:8]
    filename = f"fb-{slug}.md"
    path = FEEDBACK_DIR / filename

    # Append to existing file or create new
    mode = "a" if path.exists() else "w"
    with open(path, mode, encoding="utf-8") as f:
        if mode == "a":
            f.write("\n\n---\n\n")
        f.write(_feedback_to_markdown(req))

    # Force KB reindex on next analyze call by resetting flag
    try:
        from routes.analyze import _reset_kb
        _reset_kb()
    except Exception:
        pass

    return {
        "status": "saved",
        "diagram": req.diagram_name,
        "file": str(path.relative_to(KNOWLEDGE_DIR)),
    }

File: services/drakon-agent-flue/agents/tools/feedback.ts
import { Type, defineTool } from '@flue/runtime';

async function md5Hex(str: string): Promise<string> {
  const msgUint8 = new TextEncoder().encode(str);
  const hashBuffer = await crypto.subtle.digest('SHA-256', msgUint8);
  const hashArray = Array.from(new Uint8Array(hashBuffer));
  const hashHex = hashArray.map(b => b.toString(16).padStart(2, '0')).join('');
  return hashHex.substring(0, 8);
}

function buildFeedbackMarkdown(diagramName: string, feedbackText: string, correctedIr?: any): string {
  const ts = new Date().toISOString();
  let lines = [
    `## Feedback: ${diagramName}`,
    ``,
    `**Date:** ${ts}`,
    `**Diagram:** \`${diagramName}\``,
    ``,
    `### What was wrong`,
    `${feedbackText}`
  ];

  if (correctedIr) {
    lines = lines.concat([
      ``,
      `### Corrected IR`,
      `\`\`\`json`,
      JSON.stringify(correctedIr, null, 2),
      `\`\`\``,
      ``,
      `**Key corrections:**`
    ]);

    const items = correctedIr.items || {};
    for (const [nodeId, nodeAny] of Object.entries(items)) {
      const node = nodeAny as any;
      if (!node) continue;
      if (node.type === 'question') {
        lines.push(`- \`${nodeId}\`: question→one=${node.one}, two=${node.two}`);
      } else if (node.type === 'action') {
        lines.push(`- \`${nodeId}\`: action content=\`${(node.content || '').substring(0, 60)}\``);
      }
    }
  }

  return lines.join('\n');
}

export const feedbackTool = defineTool({
  name: 'feedback',
  description: 'Persist diagram feedback and corrections into the knowledge base.',
  parameters: Type.Object({
    diagramName: Type.String({ description: 'The name of the diagram' }),
    feedback: Type.String({ description: 'Description of the error' }),
    correctedIr: Type.Any({ description: 'Optional corrected DRAKON IR' }),
  }),
  execute: async ({ diagramName, feedback, correctedIr }, context: any) => {
    const slug = await md5Hex(diagramName + feedback);
    const filename = `fb-${slug}.md`;
    const markdown = buildFeedbackMarkdown(diagramName, feedback, correctedIr);

    // 1. Save to KV
    if (context?.env?.KNOWLEDGE_BASE) {
      try {
        await context.env.KNOWLEDGE_BASE.put(`feedback/${filename}`, markdown);
      } catch (e: any) {
        console.error('Error writing feedback to KV:', e);
      }
    }

    // 2. Save to local filesystem if process/fs are available
    let savedLocal = false;
    let localPath = '';
    if (typeof process !== 'undefined') {
      try {
        const fs = await import('fs');
        const path = await import('path');

        const possiblePaths = [
          path.join(process.cwd(), 'services', 'drakon-agent', 'knowledge'),
          path.join(process.cwd(), '..', 'drakon-agent', 'knowledge'),
          path.join(process.cwd(), 'knowledge'),
          '/data/data/com.termux/files/home/workspace/ai-drakon-scaffolder/services/drakon-agent/knowledge'
        ];

        let kbDir = '';
        for (const p of possiblePaths) {
          if (fs.existsSync(p)) {
            kbDir = p;
            break;
          }
        }

        if (kbDir) {
          const fbDir = path.join(kbDir, 'feedback');
          if (!fs.existsSync(fbDir)) {
            fs.mkdirSync(fbDir, { recursive: true });
          }
          const p = path.join(fbDir, filename);
          fs.writeFileSync(p, markdown, 'utf-8');
          savedLocal = true;
          localPath = p;
        }
      } catch (e: any) {
        console.error('Error writing feedback to filesystem:', e);
      }
    }

    return JSON.stringify({
      status: 'saved',
      diagram: diagramName,
      file: filename,
      savedLocal,
      localPath: savedLocal ? localPath : undefined
    }, null, 2);
  }
});
